commit 88bd11d1538de975e70d95bd4c1331cda709d7fc (refs/changes/11/3911/1)
Author: Nick Zivkovic <nick.zivkovic@joyent.com>
Date:   2018-05-04T20:50:47+00:00 (1 year, 5 months ago)
    
    TRITON-170 napi-ufds-watcher should retry when failing to list default VLAN

diff --git a/lib/default-fabric-setup.js b/lib/default-fabric-setup.js
index 8b5e4eb..855b172 100644
--- a/lib/default-fabric-setup.js
+++ b/lib/default-fabric-setup.js
@@ -53,29 +53,47 @@ function DefaultFabricSetupStream(opts) {
 
         var reqOptions = { headers: { 'x-request-id': _opts.requestId }};
 
-        self.napiClient.listFabricVLANs(_opts.uuid,
-            reqOptions,
-            function (err, vlans) {
+        function find() {
+            self.napiClient.listFabricVLANs(_opts.uuid,
+                reqOptions,
+                function (err, vlans) {
+                    var timeout = 10000;
 
-            if (err) {
-                return cb(err);
-            }
+                    if (err && err.statusCode >= 500 && err.statusCode < 600) {
+                        self.log.warn({err: err},
+                            'Server error finding network for user %s. ' +
+                            'Retrying in %sms',
+                            _opts.uuid, timeout);
+                        setTimeout(find, timeout);
+                        return;
+                    } else if (err) {
+                        self.log.error({err: err, user: _opts.uuid},
+                            'Error finding network for user %s',
+                            _opts.uuid);
+                        cb(err);
+                        return;
+                    }
 
-            var defaultVlan = vlans.reduce(function (acc, v) {
-                return acc || (v.name === opts.default_vlan.name ? v : null);
-            }, null);
+                    var defaultVlan = vlans.reduce(function (acc, v) {
+                        return acc || (v.name === opts.default_vlan.name ? v :
+                            null);
+                    }, null);
 
-            if (defaultVlan) {
-                self.log.info({ vlan: defaultVlan, user: _opts.uuid },
-                    'Default vlan exists for user');
-                _opts.defaultVlan = defaultVlan;
-                return cb();
-            }
+                    if (defaultVlan) {
+                        self.log.info({ vlan: defaultVlan, user: _opts.uuid },
+                            'Default vlan exists for user');
+                        _opts.defaultVlan = defaultVlan;
+                        cb();
+                        return;
+                    }
 
-            self.log.info({ user: _opts.uuid },
-                'Default vlan does not exist for user');
-            return cb();
-        });
+                    self.log.info({ user: _opts.uuid },
+                        'Default vlan does not exist for user');
+                    cb();
+                    return;
+                });
+        }
+        find();
     }
 
     function createDefaultVlan(_opts, cb) {
